File: frontend/src/api/mas.ts
import apiClient from './client'

export interface MasNodeProgress {
  key: string
  agent_id: string
  intent: string
  status: string
  attempt_count: number
}

export interface MasRunProgress {
  run_id: string
  status: string
  version: number
  nodes: MasNodeProgress[]
}

export interface MasPlanNode {
  key: string
  agent_id: string
  intent: string
  depends_on?: string[]
  input_contract?: Record<string, unknown>
  output_contract?: Record<string, unknown>
  parameters?: Record<string, unknown>
  resources?: Record<string, unknown>
  max_attempts?: number
  allow_skipped_dependencies?: boolean
}

export interface MasPlan {
  schema_version?: string
  title: string
  nodes: MasPlanNode[]
}

export interface MasRun {
  id: string
  plan_id: string
  status: string
  version: number
  context_summary: Record<string, unknown>
  created_at: string
  finished_at: string | null
}

export interface MasRunCreateRequest {
  plan: MasPlan
  context_summary?: Record<string, unknown>
  session_id?: string
  workspace_id?: string
}

export interface MasArtifact {
  id: string
  logical_name: string
  kind: string
  workspace_path: string
  media_type: string
  size_bytes: number
  sha256: string
  state: string
  version: number
  created_at: string
}

export const masApi = {
  createRun: (request: MasRunCreateRequest) =>
    apiClient.post<MasRun>('/mas/runs', request).then((response) => response.data),
  approveRun: (runId: string) =>
    apiClient.post<MasRun>(`/mas/runs/${runId}/approve`).then((response) => response.data),
  retryNode: (runId: string, nodeKey: string) =>
    apiClient.post<MasRun>(`/mas/runs/${runId}/nodes/${nodeKey}/retry`).then((response) => response.data),
  getProgress: (runId: string) =>
    apiClient.get<MasRunProgress>(`/mas/runs/${runId}/progress`).then((response) => response.data),
  listArtifacts: (runId: string) =>
    apiClient.get<MasArtifact[]>(`/mas/runs/${runId}/artifacts`).then((response) => response.data),
  downloadArtifact: async (runId: string, artifact: MasArtifact): Promise<void> => {
    const response = await apiClient.get(`/mas/runs/${runId}/artifacts/${artifact.id}/download`, {
      responseType: 'blob',
    })
    const url = URL.createObjectURL(response.data)
    const link = document.createElement('a')
    link.href = url
    link.download = artifact.logical_name || artifact.workspace_path.split('/').pop() || 'artifact'
    link.click()
    URL.revokeObjectURL(url)
  },
  streamProgress: async (
    runId: string,
    onProgress: (progress: MasRunProgress) => void,
    signal: AbortSignal,
  ): Promise<void> => {
    const token = localStorage.getItem('access_token')
    const response = await fetch(`/api/v1/mas/runs/${runId}/events`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      signal,
    })
    if (!response.ok || !response.body) throw new Error('MAS SSE connection failed')

    const reader = response.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    while (true) {
      const { done, value } = await reader.read()
      if (done) return
      buffer += decoder.decode(value, { stream: true })
      const frames = buffer.split('\n\n')
      buffer = frames.pop() || ''
      for (const frame of frames) {
        const data = frame.split('\n').find((line) => line.startsWith('data: '))?.slice(6)
        if (data && data !== '{}') onProgress(JSON.parse(data) as MasRunProgress)
      }
    }
  },
}

File: src/cygnusx/application/schemas/mas.py
"""Request and response schemas for MAS run metadata APIs."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field

from cygnusx.application.schemas.base import CygnusXBaseSchema
from cygnusx.domain.mas.models import ExecutionPlan, MASArtifact


class MASRunCreateRequest(CygnusXBaseSchema):
    plan: ExecutionPlan
    context_summary: dict[str, Any] = Field(default_factory=dict)
    session_id: UUID | None = None
    workspace_id: UUID | None = None
    project_id: str | None = Field(None, max_length=64)


class MASRunResponse(CygnusXBaseSchema):
    id: UUID
    plan_id: UUID
    status: str
    version: int
    context_summary: dict[str, Any]
    project_id: str | None = None
    created_at: datetime
    finished_at: datetime | None


class MASArtifactRegisterRequest(CygnusXBaseSchema):
    artifact: MASArtifact


class MASArtifactResponse(CygnusXBaseSchema):
    id: UUID
    logical_name: str
    kind: str
    workspace_path: str
    media_type: str
    size_bytes: int
    sha256: str
    state: str
    version: int
    project_id: str | None = None
    created_at: datetime


class MASNodeProgressResponse(CygnusXBaseSchema):
    key: str
    agent_id: str
    intent: str
    status: str
    attempt_count: int


class MASRunProgressResponse(CygnusXBaseSchema):
    run_id: UUID
    status: str
    version: int
    nodes: list[MASNodeProgressResponse]


class MASApprovalResolveRequest(CygnusXBaseSchema):
    approved: bool
    response: dict[str, Any] = Field(default_factory=dict)


class MASApprovalResponse(CygnusXBaseSchema):
    id: UUID
    node_key: str | None
    kind: str
    prompt: str
    options: list[dict[str, Any]]
    status: str
    response: dict[str, Any] | None
    created_at: datetime
